components/camel-etcd/src/test/java/org/apache/camel/component/etcd/EtcdTestSupport.java

```diff
@@ -35,12 +43,50 @@ public void process(Exchange exchange) throws Exception {
         }
     };
 
-    @Override
-    public boolean isCreateCamelContextPerClass() {
-        return false;
+    protected String getClientUri() {
+        return String.format(
+            "http://%s:%d",
+            getContainerHost(CONTAINER_NAME),
+            getContainerPort(CONTAINER_NAME, ETCD_CLIENT_PORT));
     }
 
     protected EtcdClient getClient() {
-        return new EtcdClient(URI.create("http://localhost:2379"));
+        return new EtcdClient(URI.create(getClientUri()));
+    }
+
+    @Override
+    protected CamelContext createCamelContext() throws Exception {
+        EtcdKeysComponent keys = new EtcdKeysComponent();
+        keys.setUris(getClientUri());
+
+        EtcdStatsComponent stats = new EtcdStatsComponent();
+        stats.setUris(getClientUri());
+
+        EtcdWatchComponent watch = new EtcdWatchComponent();
+        watch.setUris(getClientUri());
+
+        CamelContext context = super.createCamelContext();
+        context.addComponent("etcd-keys", keys);
+        context.addComponent("etcd-stats", stats);
+        context.addComponent("etcd-watch", watch);
+
+        return context;
+    }
+
+    @Override
+    protected GenericContainer<?> createContainer() {
+        return etcdContainer();
+    }
+
+    public static GenericContainer etcdContainer() {
+        return new GenericContainer(CONTAINER_IMAGE)
+            .withNetworkAliases(CONTAINER_NAME)
+            .withExposedPorts(ETCD_CLIENT_PORT, ETCD_PEER_PORT)
+            .waitingFor(Wait.forLogMessageContaining("etcdserver: set the initial cluster version", 1))
+            .withCommand(
+                "-name", CONTAINER_NAME + "-0",
+                "-advertise-client-urls", "http://0.0.0.0:" + ETCD_CLIENT_PORT,
```

Comment:
I think `0.0.0.0` here should be replaced with `DockerClientFactory.instance().dockerHostIpAddress()`, if that's the endpoint the client will be trying to reach at

Reply:
this seems to return `localhost"

Reply:
On your machine - yes :)

See "There is no place <...>" here https://bsideup.github.io/posts/testcontainers_fixed_ports/

Reply:
I don't use fixed port but the address the client connects to is made using something like

    container.getContainerIpAddress() + ":" + container.getMappedPort(ETCD_CLIENT_PORT)

Then, by changing the address as you suggested, the client fails to connect because of:

    connection refused: localhost/127.0.0.1:32804

Which seems to be expected as etcd inside the container won't bind to all addresses but on loopback

Reply:
Interesting!

I assumed the bind address is controlled with `-listen-client-urls` and not `-advertise-client-urls`. Or did you change both?


Also from here:
https://etcd.io/docs/v3.3.12/faq/

> advertise-client-urls and initial-advertise-peer-urls specify the addresses etcd clients or other etcd members should use to contact the etcd server. The advertise addresses must be reachable from the remote machines. Do not advertise addresses like localhost or 0.0.0.0 for a production setup since these addresses are unreachable from remote machines.

Reply:
mh, in fact it is another issue with a test, let me fix it

Reply:
@bsideup ok fixed now, thx for the suggestion